### ISSUE TITLE
`Handler#post()` does not get delayed to the next idle loop.

### ISSUE THREAD

**Issue by jhansche** [robolectric/robolectric#1306]
In general, calling `Handler#post(Runnable)` should enqueue the `Runnable` to be executed during the next idle pass of the `Looper`.  However, the `ShadowLooper` does not do that (via `org.robolectric.util.Scheduler#postDelayed()`).  Instead, it executes the `Runnable` immediately:

```
public synchronized void postDelayed(Runnable runnable, long delayMillis) {
  if ((!isConstantlyIdling && (paused || delayMillis > 0)) || Thread.currentThread() != associatedThread) {
    postedRunnables.add(new PostedRunnable(runnable, currentTime + delayMillis));
    Collections.sort(postedRunnables);
  } else {
    runnable.run();  // XXX this is the bug
  }
}

public synchronized void post(Runnable runnable) {
  postDelayed(runnable, 0);
}
```

Ordinarily it seems that this shouldn't be a big deal;  however, this causes an issue when the `FragmentManagerImpl` posts its own `mExecCommit` runnable.  FragmentManager relies on the fact that a `post()`'ed `Runnable` will not get executed immediately, and throws an exception:

```
java.lang.IllegalStateException: Recursive entry to executePendingTransactions
```

I don't think there should be any ill side effects of executing the `postedRunnables.add(PostedRunnable)` version for **all** cases, even when `delayMillis == 0`.  The Android framework also always enqueues the `Runnable` anyway, even if `delayMillis` is 0.

**Comment by kriegfrj** [comment]
Ok, a couple of comments have come up in other issues recently that trace their ultimate cause back to this issue. I'm more of a latecomer to this discussion than many of the others who have proceeded me, but I've had a fair bit of experience with the scheduling code recently so hopefully what I have to say might be useful.

@roman-mazur noted in his last comment above that:

> Note that the tests [in Robolectric] are not run in an event loop. Hence, it's rather controversial.

This is not strictly true. It would be more accurate to say that the tests are not run in an event loop *by default*, but if you pause the scheduler then they will effectively be run in an event loop equivalent. So this is more of a question of which behaviour should be the default.

In detail, there are three modes of operation for the scheduler which controls how events are dispatched (from the Javadoc https://github.com/robolectric/robolectric/blob/master/robolectric-utils/src/main/java/org/robolectric/util/Scheduler.java):
* *paused*: if paused, then no posted events will be run unless the Scheduler is explicitly instructed to do so.
* *normal*/*unpaused*: if not paused but not set to idle constantly, then the Scheduler will automatically run any `Runnable`s that are scheduled to run at or before the Scheduler's current time, but it won't automatically run any future events. To run future events the Scheduler needs to have its clock advanced.
* *idling constantly*: if `idleConstantly(true)` is called, then the Scheduler will continue looping through posted events (including future events), advancing its clock as it goes.

The current default behaviour is for the scheduler to run *unpaused*. The question raised by this ticket basically boils down to whether the default should be *paused* instead.

My argument is that no matter which we set as default, we could wind up potentially confusing a group of Robolectric users - granted that with *unpaused* as the default people may need to know to pause the scheduler to avoid certain problems, but conversely if you make *paused* the default, people will need to know to advance the scheduler for their tests to work. So if this were a "greenfield" project there'd be arguments both ways. As both @erd and I argued in #1727, the fact that there is an existing installed codebase that depends on *unpaused* being the default behaviour tips this decision (IMO) in favour if keeping the status quo.

But I think there's some good news... there might be a generic solution that might keep everyone happy. It occurred to me that we have hit (and solved) this problem before in `ActivityController` (since moved to `ComponentController`). The solution in that case was to use `ShadowLooper.runPaused()` - which makes sure that the Scheduler is paused before handing control off to the real Android code, and then when execution flow returns to Robolectric set the scheduler's "paused" state back to what it was before. If your Robolectric test is running in *unpaused* mode, then from the test's point of view you will still be getting the same effect of immediately scheduled runnables to execute immediately; however, because the scheduler is paused while the real Android code is running this ensures that these events are still queued while the code-under-test is running and not executed inline. On reflection, it seems that #1351 is another way of achieving the same end to deal with the particular case of nested posts.

So I think that the ultimate solution to this issue is to take the implementation is to take the approach of `ComponentController` implementation and somehow apply it globally across all of Robolectric. This could either be done in a case-by-case basis (eg, `ComponentController`, #1351, and #1809 could probably be fixed by doing something similar in `ShadowChoreographer`) - however, I think that a much neater & more scalable solution would put it somewhere closer to the core of Robolectric, and automatically pause the schedulers at the point were execution passes from Robolectric to the app-under-test. This might be difficult, but we might at least be able to cover 90% of cases by doing it at the boundary between Robolectric and real Android - given that this execution boundary is already instrumented by Robolectric this should be theoretically feasible at least. Such a solution would greatly simplify `ComponentController` and its two subclasses, solve #1809, and probably also make the fixes of both #1351 and #1711 redundant. The result (if we can get it to work) would be that the status quo is preserved but the out-of-order problems that can arise within real Android code will all disappear, and we can greatly simplify the scheduling code base. All we'd need to do is pinpoint the best place to put this scheduler pause/unpause (which I have not had a chance to investigate further yet - this being a bit of a high-level brainstorm).

As part of such a wide-ranging revamp, the Scheduler states need to be handled better too. There are only three valid states, but because they are represented by two booleans a fourth indeterminate state is configurable (idling constantly + paused). Arguably these states would be better represented as a three-valued `enum`. Also note that (in part because of this inconsistency) the current `runPaused()` implementation in `ComponentController` still won't work if you're configured to run constantly.

Additional thoughts/comments appreciated.

**Comment by kriegfrj** [comment]
>... however, I think that a much neater & more scalable solution would put it somewhere closer to the 
> core of Robolectric, and automatically pause the schedulers at the point were execution passes
> from  Robolectric to the app-under-test. This might be difficult, but we might at least be able to 
> cover 90% of cases by doing it at the boundary between Robolectric and real Android - given that
> this execution boundary is already instrumented by Robolectric this should be theoretically feasible 
> at least.

I've been playing around with this idea in my spare time the last few days and I think it might work. The basic idea is to put some fixes into `ShadowWrangler` (in the `Plan`s' `run()` methods) so that it increments a counter every time that an instrumented method is called and decrements it when it exits. When the counter gets to zero, it will process any queued events if the scheduler is in the `unPaused()` state. Hopefully this will not affect performance too much - a more efficient implementation might be to inline the code in the instrumentation phase instead, but I'm a bit scared of that code so I'm going to stick with this implementation as a proof-of-concept and leave the instrumentation implementation for a future optimisation if it works. I'm also not sure how it will interact with @lexs' enhancement in #1840, which I haven't looked into in detail yet.

I'll try and get a more solid implementation of this done by the weekend sometime for people to look at and comment on. In the meantime, any high-level comments from interested members of the team (@erd, @jongerrish, @charlesmunger @erichaugh) would be appreciated.

**Comment by kriegfrj** [comment]
I have finally managed to push a fix for this (as proposed in the previous post) in the PR in #2166. ShadowWrangler now wraps all calls to instrumented methods (shadowed and unshadowed) in a block()/unblock() pair on the master scheduler. This means that posts from within those methods won't run immediately but will get queued; if in the UNPAUSED mode all immediately pending tasks will be run immediately when the last call to unBlock() is made.

**Comment by ritave** [comment]
I wanted to confirm this issue still persists in Robolectric 3, found it when using ```GridViewPager``` and trying to scroll to next page.

Method proposed by @roman-mazur helps here.
```java
Robolectric.getBackgroundThreadScheduler().pause();
Robolectric.getForegroundThreadScheduler().pause();
scrollRight();
Robolectric.getBackgroundThreadScheduler().unPause();
Robolectric.getForegroundThreadScheduler().unPause();
```

[To not clutter this issue I'm posting a stacktrace on pastebin](http://pastebin.com/1n0MJB2q)

**Comment by kriegfrj** [comment]
Yes, sorry, it is fixed in #2166 but that PR still has some outstanding work and hasn't been merged into master yet.

**Comment by TWiStErRob** [comment]
@kriegfrj any news?

**Comment by kriegfrj** [comment]
@TWiStErRob (and anyone else still listening) - I'm so very sorry but Real Life(TM) has most definitely gotten in the way and I can't guess when I might be able to get back to this. It's going to be that much more difficult to merge now too because I know that there were substantial changes in some of the core code.
